packages/teams-js/test/public/mail.spec.ts

```diff
@@ -0,0 +1,274 @@
+        .openMailItem(openMailItemParams)
+        .catch(e => expect(e).toMatchObject(new Error('The library has not yet been initialized')));
+    });
+
+    Object.keys(FrameContexts)
+      .map(k => FrameContexts[k])
+      .forEach(frameContext => {
+        it(`should not allow calls from ${frameContext} context`, async () => {
+          if (frameContext === FrameContexts.content) {
+            return;
+          }
+
+          await utils.initializeWithContext(frameContext);
+
+          await mail
+            .openMailItem(openMailItemParams)
+            .catch(e =>
+              expect(e).toMatchObject(
+                new Error(
+                  `This call is only allowed in following contexts: ["content"]. Current context: "${frameContext}".`,
+                ),
+              ),
+            );
+        });
+      });
+
+    it('should not allow calls if runtime does not support mail', async () => {
+      await utils.initializeWithContext('content');
+      utils.setRuntimeConfig({ apiVersion: 1, supports: {} });
+
+      await mail.openMailItem(openMailItemParams).catch(e => expect(e).toBe('Not Supported'));
+    });
+
+    it('should throw if a null itemId is supplied', async () => {
+      await utils.initializeWithContext('content');
+      utils.setRuntimeConfig({ apiVersion: 1, supports: { mail: {} } });
+
+      await mail
+        .openMailItem({ itemId: null })
+        .catch(e => expect(e).toMatchObject(new Error('Must supply an itemId to openMailItem')));
+    });
+
+    it('should throw if an undefined itemId is supplied', async () => {
+      await utils.initializeWithContext('content');
+      utils.setRuntimeConfig({ apiVersion: 1, supports: { mail: {} } });
+
+      await mail
+        .openMailItem({ itemId: undefined })
+        .catch(e => expect(e).toMatchObject(new Error('Must supply an itemId to openMailItem')));
+    });
+
+    it('should throw if an empty itemId is supplied', async () => {
+      await utils.initializeWithContext('content');
+      utils.setRuntimeConfig({ apiVersion: 1, supports: { mail: {} } });
+
+      await mail
+        .openMailItem({ itemId: '' })
+        .catch(e => expect(e).toMatchObject(new Error('Must supply an itemId to openMailItem')));
+    });
+
+    it('should successfully throw if the openMailItem message sends and fails', async () => {
+      await utils.initializeWithContext('content');
+      utils.setRuntimeConfig({ apiVersion: 1, supports: { mail: {} } });
+
+      const openMailItemPromise = mail.openMailItem(openMailItemParams);
+
+      const openMailItemMessage = utils.findMessageByFunc('mail.openMailItem');
+
+      const data = {
+        success: false,
+        error: dataError,
+      };
+
+      utils.respondToMessage(openMailItemMessage, data.success, data.error);
+
+      let error;
+      try {
+        await openMailItemPromise;
+      } catch (e) {
+        error = e;
+      }
+
+      expect(error).toMatchObject(new Error(dataError));
```

Comment:
I was curious about why this changed and appreciate the offline explanation (there was nothing to fail if no exception was thrown). I am curious if we could use expect.assertions here: https://stackoverflow.com/questions/50816254/necessary-to-use-expect-assertions-if-youre-awaiting-any-async-function-calls

(My curiosity drove me to try it locally and the test code is actually throwing 5 assertions, not sure why)

Reply:
More good questions which led me to find more problems with our unit tests.

It turned out that the utils.ts file in the tests directory contained 4 expects statements that were running in over half of our tests and just silently passing all the time. I've converted those to be throws and then added expect.assertions(n) to any function that is checking expects of an async call in a .then or .catch. This will verify that we are hitting the right number of expect checks without making us jump through hoops like this.

Reply:
(To be clear, you could do what I had originally and jump through hoops to validate that you were actually testing what you meant but hey now you can do it easily)